File: workspace/Programming/Test/py/list/MatrixOps.py
# Program to do Matrix operations add, subtract and multiply

def MatAdd(X, Y):
    result = []
    for i in range(len(X)):
        r = []
        for j in range(len(X[i])):
            r.append(X[i][j] + Y[i][j])
        result.append(r)
    return result

def MatSub(X, Y):
    result = []
    for i in range(len(X)):
        r = []
        for j in range(len(X[i])):
            r.append(X[i][j] - Y[i][j])
        result.append(r)
        return result

def MatMul(X, Y):
    result = []
    for i in range(len(X)):
        r = []
        for j in range(len(Y[0])):
            s = 0
            for k in range(len(Y)):
                s += X[i][k] * Y[k][j]
            r.append(s)
        result.append(r)
    return result

def main():
    X = [[12,7,3], [4 ,5,6], [7 ,8,9]]
    Y = [[5,8,1], [6,7,3], [4,5,9]]
    print(X)
    print(Y)
    print(MatAdd(X, Y))
    print(MatSub(X, Y))
    print(MatMul(X, Y))


if __name__ == "__main__":
    main()

File: workspace/Programming/MatrixOps/MatrixOps.cpp
#include <bits/stdc++.h>

using namespace std;

typedef vector<int> matrixrow;
typedef vector<vector<int>> matrix;

matrix MatAdd(matrix& X, matrix& Y) {
	matrix result;

	for (size_t i = 0; i < X.size(); i++) {
		matrixrow r;
		for (size_t j = 0; j < X[i].size(); j++)
			r.push_back(X[i][j] + Y[i][j]);
		result.push_back(r);
	}

	return result;
}

matrix MatSub(matrix& X, matrix& Y) {
	matrix result;

	for (size_t i = 0; i < X.size(); i++) {
		matrixrow r;
		for (size_t j = 0; j < X[i].size(); j++)
			r.push_back(X[i][j] - Y[i][j]);
		result.push_back(r);
	}

	return result;
}

matrix MatMul(matrix& X, matrix& Y) {
	matrix result;
	for (size_t i = 0; i < X.size(); i++) {
		matrixrow row;
		for (size_t j = 0; j < Y[0].size(); j++) {
			int e = 0;
			for (size_t k = 0; k < Y.size(); k++) {
				e += X[i][k] * Y[k][j];
			}
			row.push_back(e);
		}
		result.push_back(row);
	}
	return result;
}

void printV(matrix& r) {
	for (size_t i = 0; i < r.size(); i++) {
		cout << "[ " << flush;
		for (size_t j = 0; j < r[i].size(); j++)
			cout << r[i][j] << " " << flush;
		cout << "]" << endl;
	}
}

int main()
{
	matrix X = {{12,7,3}, {4 ,5,6}, {7 ,8,9}};
	matrix Y = {{5,8,1}, {6,7,3}, {4,5,9}};
	matrix result;

	printV(X);
	printV(Y);

	result = MatAdd(X, Y);
	printV(result);

	result = MatSub(X, Y);
	printV(result);

	result = MatMul(X, Y);
	printV(result);
	
	return 0;
}
